include/hipSYCL/std/stdpar/detail/sycl_glue.hpp

```diff
@@ -171,16 +172,26 @@ class unified_shared_memory {
       return ptr;
 
     } else {
-      if(alignment != 0)
-        return ::aligned_alloc(alignment, n);
-      else
-        return ::malloc(n);
+      if(alignment != 0) {
+        void *ptr = 0;
+        posix_memalign(&ptr, alignment, n);
+        return ptr;
+      } else
+        return __libc_malloc(n);
+    }
+  }
+
+  static void memset(void* ptr, int value, std::size_t num_bytes) {
+    if(thread_local_storage::get().disabled_stack == 0) {
+      detail::single_device_dispatch::get_queue().memset(ptr, value, num_bytes);
```

Comment:
Three remarks here:
* I think here we definitely need a `wait()` call. Otherwise there's no guarantee the memset is complete by the time `calloc` returns the pointer.
* Also, it might be worth a look into whether `hipsycl::algorithms::fill()` from `algorithms/algorithm.hpp` might be a better fit here. It's not always best for performance to use `queue::memset()`, sometimes a `queue::parallel_for()` with a manual fill kernel is better. For example, on NUMA systems on CPU, `parallel_for` might better take locality into account.
* Can you double check whether the `disabled_stack` condition is sufficient here? `malloc` has more complex conditions (`disable_stack==0`, `usm_context::is_alive()` (this prevents USM operations in the shutdown phase of the application), and a check on the pointer type (I assume this we can ignore because `memset` should only be called by `calloc` in the USM path, when we know that it is a USM pointer)

Reply:
> I think here we definitely need a `wait()` call. Otherwise there's no guarantee the memset is complete by the time `calloc` returns the pointer.

Yup, that makes sense.

> Also, it might be worth a look into whether `hipsycl::algorithms::fill()` from `algorithms/algorithm.hpp` might be a better fit here. 

Yes that sounds like a better approach.

>   Can you double check whether the `disabled_stack` condition is sufficient here? `malloc` has more complex conditions (`disable_stack==0`, `usm_context::is_alive()`)

`malloc` currently does not check whether `usm_context::is_alive()` (or am I missing something?), only `free` does, that's why I didn't add a check for that here. But I think I might redo this whole part a bit differently anyway and use `malloc` to allocate memory and use either `hipsycl::algorithms::fill_n` or libc's `memset` if `malloc` gave us a pointer with `hipsycl::sycl::get_pointer_type == unknown` to zero the memory region. So all allocations are then done using `malloc` and we don't need to reinvent the wheel here.